lib/trie/trie.go

```diff
@@ -167,675 +177,888 @@ func (t *Trie) Entries() map[string][]byte {
-func (t *Trie) NextKey(key []byte) []byte {
-	k := codec.KeyLEToNibbles(key)
+// NextKey returns the next key in the trie in lexicographic order.
+// It returns nil if no next key is found.
+func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
+	prefix := []byte(nil)
+	key := codec.KeyLEToNibbles(keyLE)
 
-	next := nextKey(t.root, nil, k)
-	if next == nil {
+	nextKey := findNextKey(t.root, prefix, key)
+	if nextKey == nil {
 		return nil
 	}
 
-	return codec.NibblesToKeyLE(next)
+	nextKeyLE = codec.NibblesToKeyLE(nextKey)
+	return nextKeyLE
 }
 
-func nextKey(curr Node, prefix, key []byte) []byte {
-	switch c := curr.(type) {
-	case *node.Branch:
-		fullKey := append(prefix, c.Key...)
-		var cmp int
-		if len(key) < len(fullKey) {
-			if bytes.Compare(key, fullKey[:len(key)]) == 1 { // arg key is greater than full, return nil
-				return nil
-			}
+func findNextKey(parent Node, prefix, searchKey []byte) (nextKey []byte) {
+	if parent == nil {
+		return nil
+	}
 
-			// the key is lexicographically less than the current node key. return first key available
-			cmp = 1
-		} else {
-			// if cmp == 1, then node key is lexicographically greater than the key arg
-			cmp = bytes.Compare(fullKey, key[:len(fullKey)])
-		}
+	if parent.Type() == node.LeafType {
+		parentLeaf := parent.(*node.Leaf)
+		return findNextKeyLeaf(parentLeaf, prefix, searchKey)
+	}
 
-		// if length of key arg is less than branch key,
-		// return key of first child, or key of this branch,
-		// if it's a branch with value.
-		if (cmp == 0 && len(key) == len(fullKey)) || cmp == 1 {
-			if c.Value != nil && bytes.Compare(fullKey, key) > 0 {
-				return fullKey
-			}
+	// Branch
+	parentBranch := parent.(*node.Branch)
+	return findNextKeyBranch(parentBranch, prefix, searchKey)
+}
 
-			for i, child := range c.Children {
-				if child == nil {
-					continue
-				}
+func findNextKeyLeaf(leaf *node.Leaf, prefix, searchKey []byte) (nextKey []byte) {
+	parentLeafKey := leaf.Key
+	fullKey := make([]byte, 0, len(prefix)+len(parentLeafKey))
+	fullKey = append(fullKey, prefix...)
+	fullKey = append(fullKey, parentLeafKey...)
 
-				next := nextKey(child, append(fullKey, byte(i)), key)
-				if len(next) != 0 {
-					return next
-				}
-			}
-		}
+	if keyIsLexicographicallyBigger(searchKey, fullKey) {
+		return nil
+	}
 
-		// node key isn't greater than the arg key, continue to iterate
-		if cmp < 1 && len(key) > len(fullKey) {
-			idx := key[len(fullKey)]
-			for i, child := range c.Children[idx:] {
-				if child == nil {
-					continue
-				}
-
-				next := nextKey(child, append(fullKey, byte(i)+idx), key)
-				if len(next) != 0 {
-					return next
-				}
-			}
+	return fullKey
+}
+
+func findNextKeyBranch(parentBranch *node.Branch, prefix, searchKey []byte) (nextKey []byte) {
+	fullKey := make([]byte, 0, len(prefix)+len(parentBranch.Key))
+	fullKey = append(fullKey, prefix...)
+	fullKey = append(fullKey, parentBranch.Key...)
```

Comment:
Make a helper function for this, since similar logic is being used at a bunch of places. May be something like `sliceWithCapacity(capacity int, arrays ...[]byte)`

Reply:
Added `concatenateSlices` together with a Benchmark to explain why it's here eventually.